docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+specific design considerations. 
+
+## Discussion
+
+If this feature is to be implemented we need to clearly define the following:
+1. What does banning a peer mean:
+   1. A peer can be simply disconnected from. 
+   2. Peer is disconnected from and banned.
+   3. Conditions for the peer to be banned.
+2. If `CheckTx` signals a peer should be banned, how do we know which peer(s) to ban?
+3. Are there possible attack scenarios by allowing this. 
+
+Any further mentions of `banning` will be agnostic to the actual way banning is implemented by the p2p layer. 
+
+
+### 1. What does banning a peer mean 
+
+Tendermint recognizes that peers can accept transactions into their mempool as valid but then when the state changes, they can become invalid. There are also transactions that are received that could never have been valid (for examle due to misconfiguration on one node). 
+We thus differentiate two scenarios - a) where `CheckTx` fails due to reasons already 
+known and b) where `CheckTx` deems a transactions could never have been valid.
+
+For the sake of simplicity, in the remainder of the text we will distinguish the failures due to a) as failures 
+signaled with `ResponseCheckTx.code = 1` and the failures described in b), failures with `ResponseCheckTx.code > 1`.
+
+For a), a peer sends transactions that **repeatedly** fail CheckTx with `ResponseCheckTx.code = 1`, and is banned or disconnected from to avoid this. In this case we need to define what repeatedly means. 
+
+For b) we need to understand what is the potential reason a transaction could never have been valid on one node, but passes `CheckTx` on another node.  
+We need to understand all the possible scenarios in which this can happen:
+
+1. What happens if a node is misconfigured and allows, for example, very large transactions into the mempool. 
+This node would then gossip these transactions and they would always fail on other nodes. 
+Is this a scenario where we want nodes to disconnect from this peer and ban it without this peer being neccessarily malicious?
+2. Are all other reasons for this to happen sign of malicious behaviour where a node explicitly lies? How can `CheckTx` pass on a valid node, but fail on another valid node with a `ResponseCheckTx.code > 1`?
+If such behaviour is only possible when a peer is malicious, should this peer be punished or banned forever? Note that 
+we cannot know whether a node is a validator in order for it to be punished. Gossiping this behaviour to other peers pro-actively
+also entails a different set of problems with it - how do we know we can trust peers who tell us to ban other peers. For these reasons, understanding the actual reason for these failures can be left for future work. 
+
+For now, we will disconnect and ban the peer regardless of the exact reason a transaction 
+is considered to never be valid. 
+
+#### **Banning in case of ResponseCheckTx.code = 1**
```

Comment:
We're kind of jumping ahead to the implementation here, but I would be for introducing a new field in `ResponseCheckTx` that defaults to not banning the peer rather than modifying the semantics of the existing `Code` field.